app/Singletons/Mods.php

```diff
@@ -135,7 +135,22 @@ public function idsToBitset($ids): int
         );
     }
 
-    public function assertValidExclusivity(int $rulesetId, array $requiredIds, array $allowedIds): bool
+    public function assertValidExclusivity(int $rulesetId, array $modAcronyms): bool
```

Comment:
Much of this is copied from what is now `assertValidMultiplayerExclusivity()`. I would have liked to have one call the other but the annoying part is how `assertValidMultiplayerExclusivity()` relies on collecting all of the `$disallowedIds` to then check `$allowedIds`. I'd have to make this return a `Set` rather than a `bool` and at that point it just kinda feels like a weird thing to do.

Reply:
I think you can simplify the other one with something like

```php
$this->assert...(...);
$disallowedIds = ...;
foreach ($rIds => $rId) {
    $disallowedIds->add($rId, ...$this->mods[...][...])
}
...;
```

It's a bit of duplicated work but probably not too bad...?

Reply:
Fair, applied in https://github.com/ppy/osu-web/pull/11875/commits/7ce774bba0c64557fe1c28a03f25ce405b56ea88, I think.

I was a touch worried about `array_pop()` semantics, because it takes in the array as a reference, so I was sorta worried that calling the first assert was gonna empty the array fully rendering the following checks useless, but from what I can gather PHP has copy-on-write semantics on arrays when passing them as function parameters so it should be fine. (The tests also cover this off, which I tested by passing stuff by ref all the way down and seeing them break.)